<a href="https://colab.research.google.com/github/victorvargas1108/Modulo4_MineriaDatos/blob/master/Practica_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import nltk

In [3]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [0]:
#nltk.download()
#nltk.download('punkt')

In [0]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize 
from nltk.tokenize import RegexpTokenizer
import numpy as np    
from numpy import dot
from numpy.linalg import norm
import math

In [6]:
stopwords.fileids()

['arabic',
 'azerbaijani',
 'danish',
 'dutch',
 'english',
 'finnish',
 'french',
 'german',
 'greek',
 'hungarian',
 'indonesian',
 'italian',
 'kazakh',
 'nepali',
 'norwegian',
 'portuguese',
 'romanian',
 'russian',
 'slovene',
 'spanish',
 'swedish',
 'tajik',
 'turkish']

In [0]:
tokenizer = RegexpTokenizer(r'\w+')

In [0]:
text_1 ='#DíaDeLaMujer 8 de Marzo "Donde hay una mujer nace la esperanza, brotan las verdades, camina el sol, se desvanecen los miedos. Eres fuente de agua, de luchas, de vida, de justicia, de sueños. Mujer, ¿qué sería del mundo sin ti?". #DiaInternacionalDeLaMujer #8deMarzo2020'
text_2 ='Hoy a nivel mundial las mujeres sólo ocupan el 25 % de los escaños en los parlamentos y representan menos del 7 % de los líderes. En #Cuba representan + del % de la Asamblea Nacional, % del Consejo de Estado. La Revolución ha dignificado a la mujer cubana #DiaDeLaMujer'
text_3 ='Marxismo, socialismo, obreras y huelgas: sí, todas estas son claves para entender cómo nació el #DiaDeLaMujer #8M #8M2020 #8marzo2020'
text_4 ='FSIE considera que todos los agentes implicados colaboremos para conseguir un sistema educativo plural, diverso y dinámico que contribuya a eliminar los estereotipos y la brecha de género en campos como FP, Ciencia, Investigación y Tecnología #DiaDeLaMujer #FSIEMADRIDsemueve'
text_5 ='Eliminemos los obstáculos para asegurar el derecho de las mujeres y niñas a una vida libre de violencia #DíaDeLaMujer #GeneraciónIgualdad'
text_6 ='La embarcación partió el 21 de febrero de San Francisco, California, pero interrumpió su ruta de Hawái a México debido a que autoridades de Sacramento reportaron el deceso de una persona por Covid-19, quien estuvo a bordo del Grand Princess entre el 11 y el 21 de febrero. Días después de la llegada de esta primera expedición, además de la confirmación de una víctima mortal, se diagnosticaron cuatro casos en el estado relacionados con el Grand Princess, quienes actualmente se encuentran en cuarentena. Como resultado de estos nuevos casos de coronavirus, el gobierno de Estados Unidos y el crucero Grand Princess se vieron en la necesidad de realizar una valoración médica a quienes se embarcaron en la segunda expedición, la cual regresó el miércoles pasado a la bahía californiana'
text_7 ='Finalmente concluyó la subasta de la misteriosa Nintendo PlayStation. La cifra, aunque estratosférica, no cumplió las expectativas de los antiguos dueños. De cualquier manera, se convirtió en el objeto de videojuegos más caro de la historia. Antes de que Nintendo lanzara la 64 en 1996, trabajó con Sony en un proyecto que tenía como objetivo crear una unidad de disco óptico para la emblemática Super Nintendo Entertainment System, mejor conocida como SNES. Si bien se logró crear un prototipo, los también responsables de Mario Bros. decidieron descartar el producto debido a la piratería que rodeaba al CD-ROM en aquella época. Sony, sin embargo, no se quedó de brazos cruzados, y continuó trabajando en la propuesta hasta que finalmente se convirtió en la primera PlayStation. Salvo sorpresa, solo sobrevive una unidad del prototipo en la actualidad. En 2015, Dan Diebold encontró la consola en una caja de objetos que pertenecía a Olaf Olafsson, ex director de la división norteamericana de Sony Computer Entertainment. Resulta que el padre de Dan, Terry, había comprado una propiedad de Olafsson en una subasta; la Nintendo PlayStation, para sorpresa de los compradores, estaba guardada en ese lugar'
text_8 ='El monumento de piedra está arreglado con flores de avispa, una coqueta flor tropical común en el archipiélago de Solentiname, adonde la mañana de este viernes fueron enterradas de forma anticipada las reliquias del poeta Ernesto Cardenal, figura clave de la Teología de la Liberación. El entierro estaba programado para este sábado, pero fue adelantado para evitar otro agravio por parte del Gobierno de Daniel Ortega y la vicepresidenta Rosario Murillo, cuyos simpatizantes invadieron el martes la misa del funeral que se celebró en la Catedral Metropolitana de Managua.'
text_9 ='Ya en la última fase del modelo, en la que refiere a extender el conocimiento, la intención es que vayas más allá de la lección y puedas seguir el proceso de aprendizaje. Para ello, puedes utilizar la información para fantasear, imaginar e investigar soluciones ideales. Hacerle seguimiento a los detalles y formar asociaciones también hacen parte de este proceso creativo. Asimismo, el modelo TIM trabaja en el desarrollo de 16 habilidades creativas: curiosidad, aceptar el reto, considerar muchas alternativas, usar la fantasía, resaltar la esencia, mirar de otra forma, mantenerse consciente, utilizar el juego o el humor, ser original y mantener una mente abierta, entre otras. Una de las instituciones que cuenta con cursos que aplican este método y que se especializan en usar la creatividad como forma de crecimiento es la Unab Bogotá. Tiene, por ejemplo, el curso de ‘Creatividad y emprendimiento’ de pregrado, el curso de posgrados ‘Solución creativa de problemas’ y todos los talleres ofertados al sector externo por extensión universitaria.'
text_10 ='¿Cómo podría el mundo aprovechar la inteligencia artificial (IA) para el bien común, mientras reduce su potencial para ser una fuerza para el mal? En toda justicia, esta no es una pregunta nueva y, en últimas, puede aplicar para prácticamente cualquier tecnología de alto impacto, especialmente en la era digital. Lo que quizá resulta novedoso, o cuando menos sorprendente, es que sea el Vaticano el que se formule esta pregunta. Por instrucción del Papa Francisco, la Academia Pontifica por la Vida comenzó a estudiar el tema y, producto de este análisis, IBM y Microsoft firmaron este viernes el Rome Call for AI Ethics, un compromiso para establecer parámetros éticos en el desarrollo de inteligencia artificial. Lea también: ¿Cómo sumarle humanidad a la tecnología? La ética en el desarrollo de tecnologías es un tema cada vez más importante porque compañías de esta escala cada vez ofrecen soluciones para sistemas críticos para el funcionamiento de una sociedad: algoritmos que clasifican el desempeño de los profesores de escuelas públicas o análisis de datos para intentar predecir qué paciente puede desarrollar problemas como diabetes. En las manos equivocadas, con un diseño y una implementación incorrecta, esta información puede resultar en discriminación y otras injusticias, pero a gran escala.'
text_11 ='Si sientes que te ahogas, sólo recuerda que Cristo sabe calmar tormentas.'
text_12 ='A veces la mejor arma es una mano vacía, y un corazón dispuesto a servir con el alma.'
text_13 ='A pesar de todo lo que Dios sabe de nuestra pasado, presente y futuro, no pierde el amor y la ternura con la que nos invita a seguirle a pesar de lo que sea que hagamos o hayamos hecho. Siempre hay un espacio de reivindicación ante el arrepentimiento y el perdón que hay en Él.'
text_14 ='La oración es vital para convertir una Crisis en testimonio. La oración es un catalizador que coloca al espíritu sobre la materia.'
text_15 ='Lo que guardas en tu corazón condicionará tu camino y tus acciones. En tus latidos están tus pensamientos y en ellos tus pasos. Si en tu corazón no está Dios en primer lugar, lo demás no valdrá la pena.'

In [9]:
textos = [text_1,text_2,text_3,text_4,text_5,text_6,text_7,text_8,text_9,text_10,text_11,text_12,text_13,text_14,text_15]
textos_wostw=[]
x=0
for element in textos:
  t = tokenizer.tokenize(element.lower())
  twostw = []
  for word in t:  
    if word not in stopwords.words('spanish'): twostw.append(word)
  #print(twostw)
  #print(len(twostw))
  textos_wostw.append(twostw)
  print(x,textos_wostw[x])
  x = x + 1

0 ['díadelamujer', '8', 'marzo', 'mujer', 'nace', 'esperanza', 'brotan', 'verdades', 'camina', 'sol', 'desvanecen', 'miedos', 'fuente', 'agua', 'luchas', 'vida', 'justicia', 'sueños', 'mujer', 'mundo', 'diainternacionaldelamujer', '8demarzo2020']
1 ['hoy', 'nivel', 'mundial', 'mujeres', 'sólo', 'ocupan', '25', 'escaños', 'parlamentos', 'representan', 'menos', '7', 'líderes', 'cuba', 'representan', 'asamblea', 'nacional', 'consejo', 'revolución', 'dignificado', 'mujer', 'cubana', 'diadelamujer']
2 ['marxismo', 'socialismo', 'obreras', 'huelgas', 'todas', 'claves', 'entender', 'cómo', 'nació', 'diadelamujer', '8m', '8m2020', '8marzo2020']
3 ['fsie', 'considera', 'agentes', 'implicados', 'colaboremos', 'conseguir', 'sistema', 'educativo', 'plural', 'diverso', 'dinámico', 'contribuya', 'eliminar', 'estereotipos', 'brecha', 'género', 'campos', 'fp', 'ciencia', 'investigación', 'tecnología', 'diadelamujer', 'fsiemadridsemueve']
4 ['eliminemos', 'obstáculos', 'asegurar', 'derecho', 'mujeres',

In [0]:
c=0
dicc_texts ={}
for element in textos_wostw:
  tt="text" + str(c)
  #print(tt)
  dicc_texts.update ({tt: element})
  c = c + 1
#dicc_texts

In [11]:
dicc_termns = {}
for text in dicc_texts:
    for word in dicc_texts[text]:
        
#        print("EVALUAR:", word, "EN", text)
        
        if(word in dicc_termns): #incrementar palabras al diccionario
            dicc_termns[word] = dicc_termns[word] + 1
            
#            print(word, "IN", "dicc_termns")
            
        elif(word not in dicc_termns) :#agregar palabras al diccionario        
            dicc_termns[word] = 1
            
#            print(word, "NOT IN", "dicc_termns")            

print(len(dicc_termns))
dicc_termns

491


{'11': 1,
 '16': 1,
 '19': 1,
 '1996': 1,
 '2015': 1,
 '21': 2,
 '25': 1,
 '64': 1,
 '7': 1,
 '8': 1,
 '8demarzo2020': 1,
 '8m': 1,
 '8m2020': 1,
 '8marzo2020': 1,
 'abierta': 1,
 'academia': 1,
 'acciones': 1,
 'aceptar': 1,
 'actualidad': 1,
 'actualmente': 1,
 'adelantado': 1,
 'además': 1,
 'adonde': 1,
 'agentes': 1,
 'agravio': 1,
 'agua': 1,
 'ahogas': 1,
 'ai': 1,
 'algoritmos': 1,
 'allá': 1,
 'alma': 1,
 'alternativas': 1,
 'alto': 1,
 'amor': 1,
 'anticipada': 1,
 'antiguos': 1,
 'análisis': 2,
 'aplican': 1,
 'aplicar': 1,
 'aprendizaje': 1,
 'aprovechar': 1,
 'aquella': 1,
 'archipiélago': 1,
 'arma': 1,
 'arreglado': 1,
 'arrepentimiento': 1,
 'artificial': 2,
 'asamblea': 1,
 'asegurar': 1,
 'asimismo': 1,
 'asociaciones': 1,
 'aunque': 1,
 'autoridades': 1,
 'avispa': 1,
 'bahía': 1,
 'bien': 2,
 'bogotá': 1,
 'bordo': 1,
 'brazos': 1,
 'brecha': 1,
 'bros': 1,
 'brotan': 1,
 'cada': 2,
 'caja': 1,
 'california': 1,
 'californiana': 1,
 'call': 1,
 'calmar': 1,
 'camina

In [0]:
matrix = np.zeros((len(dicc_texts), len(dicc_termns)))

In [13]:
i = 0
j = 0

for word_termns in dicc_termns: #dicc_termns todos los términos
#    print()
    for word_texts in dicc_texts: #dicc_texts todos los textos
#        print("EVALUAR:", word_termns, "EN: ", word_texts)
        if(word_termns in dicc_texts[word_texts]): #si está
            print(word_termns, "IN", word_texts)
            matrix[j, i] = 1
        elif(word_termns not in dicc_texts[word_texts]): # si no está
            print(word_termns, "NOT IN", word_texts)
            matrix[j, i] = 0
        #print("se agregó: ", matrix[j,i], "en: ", j, i)
        j = j + 1
    j = 0
    i = i + 1

Se han truncado las últimas 5000 líneas del flujo de salida.
objeto NOT IN text11
objeto NOT IN text12
objeto NOT IN text13
objeto NOT IN text14
videojuegos NOT IN text0
videojuegos NOT IN text1
videojuegos NOT IN text2
videojuegos NOT IN text3
videojuegos NOT IN text4
videojuegos NOT IN text5
videojuegos IN text6
videojuegos NOT IN text7
videojuegos NOT IN text8
videojuegos NOT IN text9
videojuegos NOT IN text10
videojuegos NOT IN text11
videojuegos NOT IN text12
videojuegos NOT IN text13
videojuegos NOT IN text14
caro NOT IN text0
caro NOT IN text1
caro NOT IN text2
caro NOT IN text3
caro NOT IN text4
caro NOT IN text5
caro IN text6
caro NOT IN text7
caro NOT IN text8
caro NOT IN text9
caro NOT IN text10
caro NOT IN text11
caro NOT IN text12
caro NOT IN text13
caro NOT IN text14
historia NOT IN text0
historia NOT IN text1
historia NOT IN text2
historia NOT IN text3
historia NOT IN text4
historia NOT IN text5
historia IN text6
historia NOT IN text7
historia NOT IN text8
historia NOT I

In [16]:
matrix

array([[1., 1., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 1., 1., 1.]])

In [15]:
matrix.shape

(15, 491)

In [20]:
i=0
j=1
cc=0
while i < (len(matrix)):
  #print(i)
  while j < (len(matrix)):  
    RES = dot(matrix[i],matrix[j]) / (norm(matrix[i])*norm(matrix[j]))
    print("Texto ",i," vs Texto ",j," similitud = ",RES)
    Dist_euc=distance = math.sqrt(sum([(a - b) ** 2 for a, b in zip(matrix[i], matrix[j])]))
    print("Euclidean distance from ",i," to ",j,":",distance)
    #print (i,j, RES)
    #print(bin_cos[i,j])
    #print(i);print(J)
    cc += 1  
    j += 1
  i += 1
  j = i+1

Texto  0  vs Texto  1  similitud =  0.046524210519923545
Euclidean distance from  0  to  1 : 6.4031242374328485
Texto  0  vs Texto  2  similitud =  0.0
Euclidean distance from  0  to  2 : 5.830951894845301
Texto  0  vs Texto  3  similitud =  0.0
Euclidean distance from  0  to  3 : 6.6332495807108
Texto  0  vs Texto  4  similitud =  0.1315903389919538
Euclidean distance from  0  to  4 : 5.291502622129181
Texto  0  vs Texto  5  similitud =  0.0
Euclidean distance from  0  to  5 : 8.831760866327848
Texto  0  vs Texto  6  similitud =  0.0
Euclidean distance from  0  to  6 : 10.583005244258363
Texto  0  vs Texto  7  similitud =  0.0
Euclidean distance from  0  to  7 : 8.306623862918075
Texto  0  vs Texto  8  similitud =  0.0
Euclidean distance from  0  to  8 : 10.198039027185569
Texto  0  vs Texto  9  similitud =  0.06546536707079773
Euclidean distance from  0  to  9 : 10.723805294763608
Texto  0  vs Texto  10  similitud =  0.0
Euclidean distance from  0  to  10 : 5.477225575051661
Texto  0

**Proceso por frecuencias**

In [0]:
matrix = np.zeros((len(dicc_texts), len(dicc_termns)))

In [22]:
i = 0
j = 0
for word_termns in dicc_termns: 
    for word_texts in dicc_texts: 
        if(word_termns in dicc_texts[word_texts]): #si está
            print(word_termns, "IN", word_texts)
            matrix[j, i] = dicc_termns[word_termns]
        elif(word_termns not in dicc_texts[word_texts]): # si no está
            print(word_termns, "NOT IN", word_texts)
            matrix[j, i] = 0
        print("se agregó: ", matrix[j,i], "en: ", j, i)
        j = j + 1        
    j = 0
    i = i + 1

Se han truncado las últimas 5000 líneas del flujo de salida.
se agregó:  0.0 en:  5 324
mirar NOT IN text6
se agregó:  0.0 en:  6 324
mirar NOT IN text7
se agregó:  0.0 en:  7 324
mirar IN text8
se agregó:  1.0 en:  8 324
mirar NOT IN text9
se agregó:  0.0 en:  9 324
mirar NOT IN text10
se agregó:  0.0 en:  10 324
mirar NOT IN text11
se agregó:  0.0 en:  11 324
mirar NOT IN text12
se agregó:  0.0 en:  12 324
mirar NOT IN text13
se agregó:  0.0 en:  13 324
mirar NOT IN text14
se agregó:  0.0 en:  14 324
mantenerse NOT IN text0
se agregó:  0.0 en:  0 325
mantenerse NOT IN text1
se agregó:  0.0 en:  1 325
mantenerse NOT IN text2
se agregó:  0.0 en:  2 325
mantenerse NOT IN text3
se agregó:  0.0 en:  3 325
mantenerse NOT IN text4
se agregó:  0.0 en:  4 325
mantenerse NOT IN text5
se agregó:  0.0 en:  5 325
mantenerse NOT IN text6
se agregó:  0.0 en:  6 325
mantenerse NOT IN text7
se agregó:  0.0 en:  7 325
mantenerse IN text8
se agregó:  1.0 en:  8 325
mantenerse NOT IN text9
se agregó:  0

In [23]:
matrix

array([[2., 1., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 1., 1., 1.]])

In [24]:
matrix.shape

(15, 491)

In [25]:
i=0
j=1
cc=0
bin_cos=(len(matrix)-1,len(matrix)-1)
while i < (len(matrix)):
  #print(i)
  while j < (len(matrix)):  
    RES = dot(matrix[i],matrix[j]) / (norm(matrix[i])*norm(matrix[j]))
    print("Texto ",i," vs Texto ",j," similitud = ",RES)
    Dist_euc=distance = math.sqrt(sum([(a - b) ** 2 for a, b in zip(matrix[i], matrix[j])]))
    print("Euclidean distance from ",i," to ",j,":",distance)
    #print (i,j, RES)
    #print(bin_cos[i,j])
    #print(i);print(J)
    cc += 1  
    j += 1
  i += 1
  j = i+1

Texto  0  vs Texto  1  similitud =  0.18766297265136728
Euclidean distance from  0  to  1 : 8.831760866327848
Texto  0  vs Texto  2  similitud =  0.0
Euclidean distance from  0  to  2 : 8.660254037844387
Texto  0  vs Texto  3  similitud =  0.0
Euclidean distance from  0  to  3 : 9.219544457292887
Texto  0  vs Texto  4  similitud =  0.38334908600273254
Euclidean distance from  0  to  4 : 6.708203932499369
Texto  0  vs Texto  5  similitud =  0.0
Euclidean distance from  0  to  5 : 12.083045973594572
Texto  0  vs Texto  6  similitud =  0.0
Euclidean distance from  0  to  6 : 15.165750888103101
Texto  0  vs Texto  7  similitud =  0.0
Euclidean distance from  0  to  7 : 10.677078252031311
Texto  0  vs Texto  8  similitud =  0.0
Euclidean distance from  0  to  8 : 13.341664064126334
Texto  0  vs Texto  9  similitud =  0.17463705352625183
Euclidean distance from  0  to  9 : 14.7648230602334
Texto  0  vs Texto  10  similitud =  0.0
Euclidean distance from  0  to  10 : 8.306623862918075
Texto  

La mejor texto es Texto  0  vs Texto  4  similitud =  0.38334908600273254 y con una distancia Euclidiana 6.708203932499369, es la mejor opción.